payroll/search.py

```
@@ -154,18 +154,107 @@ def search(self, params, pagesize, **extra_kwargs):
+
+        total_hits = sum(hits for _, hits in entity_edges)
+
+        search_results = self._composite_search(entity_edges,
+                                                pagesize,
+                                                offset,
+                                                query_string,
+                                                **extra_kwargs)
+
+        return LazyPaginatedResults(search_results, total_hits)
+
+    def _composite_search(self,
+                          entity_edges,
+                          pagesize,
+                          offset,
+                          query_string,
+                          **extra_kwargs):
+        '''
+        :entity_edges - list of tuples, containing entity type and number of
+            hits for the given search term
+        :pagesize - number of search results to return
+        :offset - (current page - 1) * offset, i.e., the starting index of the
+            current page of results
+        :query_string - Solr query string
+
+        Given the number of hits per entity type, generate on object containing
+        `pagesize` search results. This object contains multiple types of
+        entity when the number of hits for a given entity type is not divisible
+        by `pagesize`, i.e., the last page of its results contains fewer than
+        `pagesize` items.
+
+        For example, if there are three unit results and 10 department results,
+        and the requested page size is five, this method would return a list of
+        three units and two departments, for a total of five results.
+        '''
+        upper_bound = 0
+        lower_bound = 0
+
+        search_results = []
+        entity_type_idx = 0
+
+        while len(search_results) < pagesize:
+            try:
+                entity_type, entity_hits = entity_edges[entity_type_idx]
+            except IndexError:  # No more entity types, i.e., last page
+                break
```

this shouldn't happen, maybe we want to bubble this error up?

---

it can and does happen when the last page has fewer than pagesize results, like 7 units with a pagesize of 5 would have one page with 5 results and one with 2. django pagination is smart about requesting slices of the right size, so it works fine with our lazy results class.

---

i'm a bit confused. if django pagination is smart about requesting slices of the right size, doesn't that mean this shouldn't happen?

---

if django pagination is smart, then how shouldn't this prevent this from happening? I might need you to talk me through it.

---

we are not interacting with django pagination yet. the pagesize argument comes from us: it's passed as an argument to `search` from the view code, and handed through to `_composite_search`.

i meant that, once we create our object and pass it to the lazy results object, django is smart about the slice, i.e., [our sanity check](https://github.com/datamade/bga-payroll/blob/1619d90e576ae12a911ec2ba5bc578ce0e4ea1b1/payroll/search.py#L103-L111) passes.

---

ok